src/client/mesh_generator_thread.cpp

```diff
@@ -264,18 +264,61 @@ void MeshUpdateQueue::cleanupCache()
 }
 
 /*
-	MeshUpdateThread
+	MeshUpdateWorkerThread
 */
 
-MeshUpdateThread::MeshUpdateThread(Client *client):
-	UpdateThread("Mesh"),
-	m_queue_in(client)
+MeshUpdateWorkerThread::MeshUpdateWorkerThread(MeshUpdateQueue *queue_in, MutexedQueue<MeshUpdateResult> *queue_out, v3s16 *camera_offset) :
+		UpdateThread("Mesh"), m_queue_in(queue_in), m_queue_out(queue_out), m_camera_offset(camera_offset)
 {
 	m_generation_interval = g_settings->getU16("mesh_generation_interval");
 	m_generation_interval = rangelim(m_generation_interval, 0, 50);
 }
 
-void MeshUpdateThread::updateBlock(Map *map, v3s16 p, bool ack_block_to_server,
+void MeshUpdateWorkerThread::doUpdate()
+{
+	QueuedMeshUpdate *q;
+	while ((q = m_queue_in->pop())) {
+		if (m_generation_interval)
+			sleep_ms(m_generation_interval);
+		ScopeProfiler sp(g_profiler, "Client: Mesh making (sum)");
+
+		MapBlockMesh *mesh_new = new MapBlockMesh(q->data, *m_camera_offset);
+
+		
+
+		MeshUpdateResult r;
+		r.p = q->p;
+		r.mesh = mesh_new;
+		r.ack_block_to_server = q->ack_block_to_server;
+		r.urgent = q->urgent;
+
+		m_queue_out->push_back(r);
+
+		delete q;
+	}
+}
+
+/*
+	MeshUpdateManager
+*/
+
+MeshUpdateManager::MeshUpdateManager(Client *client):
+	m_queue_in(client)
+{
+	int number_of_threads = rangelim(g_settings->getS32("mesh_generation_threads"), 0, 32);
+
+	// Automatically use 33% of the system cores for mesh generation
+	if (number_of_threads == 0)
+		number_of_threads = std::thread::hardware_concurrency() / 3;
```

Comment:
`Thread::getNumberOfProcessors()`

also this should have a maximum during auto-detection
more than 4 mesh threads probably don't make sense?

Reply:
Agree, but I would cap it at 8.

Reply:
FWIW on my machine (8 cores, 16 HTs) and my scene I see diminishing return around 3 or 4 threads.

Reply:
It's 4, and the maximum if reduced from 32 to 8, just in case someone needs it.